File: Review-09-Java-Shapes/Example-6/src/main/java/edu/odu/cs/cs330/examples/shapes/ShapeFactory.java
// Thomas Kennedy
// CS 330 Fall 2020

package edu.odu.cs.cs330.examples.shapes;

/**
 * The Shape Creating Wizard.
 *
 * @author Thomas J Kennedy
 */
public class ShapeFactory {
    /**
     * Listing of known shapes.
     */
    private static final Shape[] _KNOWN_SHAPES = new Shape[] {
        new Triangle(),
        new RightTriangle(),
        new EquilateralTriangle(),
        new Square(),
        new Circle()
    };

    /**
     *  Create a Shape.
     *
     *  @param name the shape to be created
     *
     *  @return A shape with the specified name
     *      or null if no matching shape is found
     *
     * @throws CloneNotSupportedException if a model shape is found and the
     *     copy (i.e., clone) operation fails
     */
    public static Shape createShape(String name)
        throws CloneNotSupportedException
    {
        for (Shape shape : _KNOWN_SHAPES) {
            if (shape.name().equals(name)) {
                return (Shape) shape.clone();
            }
        }

        return null;
    }

    /**
     * Determine whether a given shape is known.
     *
     * @param name the shape for which to query
     *
     * @return true if the specified name corresponds to a known shape
     */
    public static boolean isKnown(String name)
    {
        for (Shape shape : _KNOWN_SHAPES) {
            if (shape.name().equals(name)) {
                return true;
            }
        }

        return false;
    }

    /**
     *  Print a list of known Shapes.
     *
     * @return a string containing a newline delimited list of known shape names
     */
    public static String listKnown()
    {
        StringBuilder bld = new StringBuilder();

        for (Shape shape : _KNOWN_SHAPES) {
            bld.append(shape.name() + "\n");
        }

        return bld.toString();
    }

    /**
     *  Determine the number of known Shapes.
     *
     *  @return the number of known shapes
     */
    public static int numberKnown()
    {
        return _KNOWN_SHAPES.length;
    }
}

File: 06-Python-Shapes/Example-7-Initial/shapes/shape_factory.py
"""
This module provides factory utilities for creating shapes. This includes
recording which Shape types are available.

"""

from __future__ import annotations

import copy

from shapes.shape import Shape
from shapes.circle import Circle
from shapes.square import Square
from shapes.triangle import (Triangle, RightTriangle, EquilateralTriangle)


_KNOWN_SHAPES = {
    "Triangle": Triangle,
    "Right Triangle": RightTriangle,
    "Equilateral Triangle": EquilateralTriangle,
    "Square": Square,
    "Circle": Circle
}  # _Dictionary_ of known shapes


def create(name: str) -> Shape:
    """
    Create a Shape

    Args:
        name: the shape to be created

    Returns:
        A shape with the specified name or null if no matching shape is found
    """

    if name in _KNOWN_SHAPES:
        return _KNOWN_SHAPES[name]()

    return None


def create_from_dimensions(name: str, values: list[float]) -> Shape:
    """
    Create a Shape

    Args:
        name: the shape to be created

        values: dictionary of values corresponding to the data needed
            to inialize a shape

    Returns:
        A shape with the specified name or null if no matching shape is found
    """

    if name in _KNOWN_SHAPES:
        return _KNOWN_SHAPES[name](*values)

    return None


def is_known(name: str) -> bool:
    """
    Determine whether a given shape is known

    Args:
        name: the shape for which to query
    """

    return name in _KNOWN_SHAPES


def list_known() -> str:
    """
    Print a list of known Shapes
    """
    return "\n".join((f"  {name:}" for name in _KNOWN_SHAPES))


def number_known() -> int:
    """
    Determine the number of known Shapes
    """

    return len(_KNOWN_SHAPES)
